File: ts/indicators/george-lane/stochastic/stochastic.ts
import { buildMetadata } from '../../core/build-metadata';
import { Bar } from '../../../entities/bar';
import { Quote } from '../../../entities/quote';
import { Scalar } from '../../../entities/scalar';
import { Trade } from '../../../entities/trade';
import { Indicator } from '../../core/indicator';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { ExponentialMovingAverage } from '../../common/exponential-moving-average/exponential-moving-average';
import { SimpleMovingAverage } from '../../common/simple-moving-average/simple-moving-average';
import { StochasticParams, MovingAverageType } from './params';

/** Interface for an indicator that accepts a scalar and returns a value. */
interface LineUpdater {
  update(sample: number): number;
  isPrimed(): boolean;
}

/** Passthrough smoother for period of 1. */
class Passthrough implements LineUpdater {
  update(v: number): number { return v; }
  isPrimed(): boolean { return true; }
}

/** Creates a moving average line updater. */
function createMA(maType: MovingAverageType | undefined, length: number, firstIsAverage: boolean): [LineUpdater, string] {
  if (length < 2) {
    return [new Passthrough(), 'SMA'];
  }

  if (maType === MovingAverageType.EMA) {
    return [new ExponentialMovingAverage({ length, firstIsAverage }), 'EMA'];
  }

  return [new SimpleMovingAverage({ length }), 'SMA'];
}

/** Function to calculate mnemonic of a __Stochastic__ indicator. */
export const stochasticMnemonic = (params: StochasticParams): string => {
  const slowKLabel = params.slowKMAType === MovingAverageType.EMA ? 'EMA' : 'SMA';
  const slowDLabel = params.slowDMAType === MovingAverageType.EMA ? 'EMA' : 'SMA';
  return `stoch(${params.fastKLength}/${slowKLabel}${params.slowKLength}/${slowDLabel}${params.slowDLength})`;
};

/**
 * George Lane's Stochastic Oscillator.
 *
 * The Stochastic Oscillator measures the position of the close relative to the
 * high-low range over a lookback period. It produces three outputs:
 *   - Fast-K: the raw stochastic value
 *   - Slow-K: a moving average of Fast-K (also known as Fast-D)
 *   - Slow-D: a moving average of Slow-K
 *
 * The indicator requires bar data (high, low, close). For scalar, quote, and
 * trade updates, the single value substitutes for all three.
 *
 * Reference:
 *
 * Lane, George C. (1984). "Lane's Stochastics". Technical Analysis of Stocks & Commodities.
 */
export class Stochastic implements Indicator {

  private readonly fastKLength_: number;

  private readonly highBuf: Float64Array;
  private readonly lowBuf: Float64Array;
  private bufferIndex = 0;
  private count = 0;

  private readonly slowKMA: LineUpdater;
  private readonly slowDMA: LineUpdater;

  private fastK_ = NaN;
  private slowK_ = NaN;
  private slowD_ = NaN;
  private primed_ = false;

  private readonly mnemonic_: string;
  private readonly description_: string;

  constructor(params: StochasticParams) {
    const fastKLength = Math.floor(params.fastKLength);
    const slowKLength = Math.floor(params.slowKLength);
    const slowDLength = Math.floor(params.slowDLength);

    if (fastKLength < 1) {
      throw new Error('fast K length should be greater than 0');
    }

    if (slowKLength < 1) {
      throw new Error('slow K length should be greater than 0');
    }

    if (slowDLength < 1) {
      throw new Error('slow D length should be greater than 0');
    }

    this.fastKLength_ = fastKLength;
    this.highBuf = new Float64Array(fastKLength);
    this.lowBuf = new Float64Array(fastKLength);

    const fia = params.firstIsAverage ?? false;
    [this.slowKMA] = createMA(params.slowKMAType, slowKLength, fia);
    [this.slowDMA] = createMA(params.slowDMAType, slowDLength, fia);

    this.mnemonic_ = stochasticMnemonic(params);
    this.description_ = 'Stochastic Oscillator ' + this.mnemonic_;
  }

  /** Indicates whether the indicator is primed. */
  public isPrimed(): boolean {
    return this.primed_;
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return buildMetadata(
      IndicatorIdentifier.Stochastic,
      this.mnemonic_,
      this.description_,
      [
        { mnemonic: this.mnemonic_ + ' fastK', description: this.description_ + ' Fast-K' },
        { mnemonic: this.mnemonic_ + ' slowK', description: this.description_ + ' Slow-K' },
        { mnemonic: this.mnemonic_ + ' slowD', description: this.description_ + ' Slow-D' },
      ],
    );
  }

  /** Updates the indicator given the next bar's close, high, and low values. Returns [FastK, SlowK, SlowD]. */
  public update(close: number, high: number, low: number): [number, number, number] {
    if (isNaN(close) || isNaN(high) || isNaN(low)) {
      return [NaN, NaN, NaN];
    }

    // Store high and low in circular buffer.
    this.highBuf[this.bufferIndex] = high;
    this.lowBuf[this.bufferIndex] = low;
    this.bufferIndex = (this.bufferIndex + 1) % this.fastKLength_;
    this.count++;

    // Need at least fastKLength bars.
    if (this.count < this.fastKLength_) {
      return [this.fastK_, this.slowK_, this.slowD_];
    }

    // Find highest high and lowest low in the window.
    let hh = this.highBuf[0];
    let ll = this.lowBuf[0];

    for (let i = 1; i < this.fastKLength_; i++) {
      if (this.highBuf[i] > hh) {
        hh = this.highBuf[i];
      }

      if (this.lowBuf[i] < ll) {
        ll = this.lowBuf[i];
      }
    }

    // Calculate Fast-K.
    const diff = hh - ll;
    if (diff > 0) {
      this.fastK_ = 100 * (close - ll) / diff;
    } else {
      this.fastK_ = 0;
    }

    // Feed Fast-K to Slow-K smoother.
    this.slowK_ = this.slowKMA.update(this.fastK_);

    // Feed Slow-K to Slow-D smoother (only when Slow-K MA is primed).
    if (this.slowKMA.isPrimed()) {
      this.slowD_ = this.slowDMA.update(this.slowK_);

      if (!this.primed_ && this.slowDMA.isPrimed()) {
        this.primed_ = true;
      }
    }

    return [this.fastK_, this.slowK_, this.slowD_];
  }

  /** Updates the indicator given the next scalar sample. */
  public updateScalar(sample: Scalar): IndicatorOutput {
    const v = sample.value;
    const [fastK, slowK, slowD] = this.update(v, v, v);
    const s1 = new Scalar();
    s1.time = sample.time;
    s1.value = fastK;
    const s2 = new Scalar();
    s2.time = sample.time;
    s2.value = slowK;
    const s3 = new Scalar();
    s3.time = sample.time;
    s3.value = slowD;
    return [s1, s2, s3];
  }

  /** Updates the indicator given the next bar sample. */
  public updateBar(sample: Bar): IndicatorOutput {
    const [fastK, slowK, slowD] = this.update(sample.close, sample.high, sample.low);
    const s1 = new Scalar();
    s1.time = sample.time;
    s1.value = fastK;
    const s2 = new Scalar();
    s2.time = sample.time;
    s2.value = slowK;
    const s3 = new Scalar();
    s3.time = sample.time;
    s3.value = slowD;
    return [s1, s2, s3];
  }

  /** Updates the indicator given the next quote sample. */
  public updateQuote(sample: Quote): IndicatorOutput {
    const v = (sample.bid + sample.ask) / 2;
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = v;
    return this.updateScalar(scalar);
  }

  /** Updates the indicator given the next trade sample. */
  public updateTrade(sample: Trade): IndicatorOutput {
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = sample.price;
    return this.updateScalar(scalar);
  }
}

File: go/indicators/georgelane/stochastic/stochastic.go
package stochastic

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/common/exponentialmovingaverage"
	"zpano/indicators/common/simplemovingaverage"
	"zpano/indicators/core"
)

// lineUpdater is an interface for indicators that accept a single scalar and return a value.
type lineUpdater interface {
	Update(float64) float64
	IsPrimed() bool
}

// passthrough is a no-op smoother for period of 1.
type passthrough struct{}

// Update returns v unchanged (no smoothing).
func (p *passthrough) Update(v float64) float64 { return v }

// IsPrimed always reports true; a passthrough has no warmup.
func (p *passthrough) IsPrimed() bool { return true }

// Stochastic is George Lane's Stochastic Oscillator.
//
// The Stochastic Oscillator measures the position of the close relative to the
// high-low range over a lookback period. It produces three outputs:
//   - Fast-K: the raw stochastic value
//   - Slow-K: a moving average of Fast-K (also known as Fast-D)
//   - Slow-D: a moving average of Slow-K
//
// The indicator requires bar data (high, low, close). For scalar, quote, and
// trade updates, the single value substitutes for all three.
//
// Reference:
//
// Lane, George C. (1984). "Lane's Stochastics". Technical Analysis of Stocks & Commodities.
type Stochastic struct {
	mu sync.RWMutex

	fastKLength int

	// Circular buffers for high and low values (size = fastKLength).
	highBuf     []float64
	lowBuf      []float64
	bufferIndex int
	count       int

	slowKMA lineUpdater
	slowDMA lineUpdater

	fastK  float64
	slowK  float64
	slowD  float64
	primed bool

	mnemonic string
}

// NewStochastic returns an instance of the indicator created using supplied parameters.
func NewStochastic(p *StochasticParams) (*Stochastic, error) {
	const (
		invalid   = "invalid stochastic parameters"
		fmts      = "%s: %s"
		fmtw      = "%s: %w"
		minLength = 1
	)

	if p.FastKLength < minLength {
		return nil, fmt.Errorf(fmts, invalid, "fast K length should be greater than 0")
	}

	if p.SlowKLength < minLength {
		return nil, fmt.Errorf(fmts, invalid, "slow K length should be greater than 0")
	}

	if p.SlowDLength < minLength {
		return nil, fmt.Errorf(fmts, invalid, "slow D length should be greater than 0")
	}

	// Create Slow-K smoother.
	slowKMA, slowKLabel, err := createMA(p.SlowKMAType, p.SlowKLength, p.FirstIsAverage)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	// Create Slow-D smoother.
	slowDMA, slowDLabel, err := createMA(p.SlowDMAType, p.SlowDLength, p.FirstIsAverage)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	mnemonic := fmt.Sprintf("stoch(%d/%s%d/%s%d)", p.FastKLength,
		slowKLabel, p.SlowKLength, slowDLabel, p.SlowDLength)

	return &Stochastic{
		fastKLength: p.FastKLength,
		highBuf:     make([]float64, p.FastKLength),
		lowBuf:      make([]float64, p.FastKLength),
		slowKMA:     slowKMA,
		slowDMA:     slowDMA,
		fastK:       math.NaN(),
		slowK:       math.NaN(),
		slowD:       math.NaN(),
		mnemonic:    mnemonic,
	}, nil
}

func createMA(maType MovingAverageType, length int, firstIsAverage bool) (lineUpdater, string, error) {
	if length < 2 { //nolint:mnd
		return &passthrough{}, "SMA", nil
	}

	switch maType {
	case EMA:
		ema, e := exponentialmovingaverage.NewExponentialMovingAverageLength(
			&exponentialmovingaverage.ExponentialMovingAverageLengthParams{
				Length:         length,
				FirstIsAverage: firstIsAverage,
			})
		if e != nil {
			return nil, "", e
		}

		return ema, "EMA", nil
	default:
		sma, e := simplemovingaverage.NewSimpleMovingAverage(
			&simplemovingaverage.SimpleMovingAverageParams{Length: length})
		if e != nil {
			return nil, "", e
		}

		return sma, "SMA", nil
	}
}

// IsPrimed indicates whether the indicator is primed.
func (s *Stochastic) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes the output data of the indicator.
func (s *Stochastic) Metadata() core.Metadata {
	desc := "Stochastic Oscillator " + s.mnemonic

	return core.BuildMetadata(
		core.Stochastic,
		s.mnemonic,
		desc,
		[]core.OutputText{
			{Mnemonic: s.mnemonic + " fastK", Description: desc + " Fast-K"},
			{Mnemonic: s.mnemonic + " slowK", Description: desc + " Slow-K"},
			{Mnemonic: s.mnemonic + " slowD", Description: desc + " Slow-D"},
		},
	)
}

// Update updates the indicator given the next bar's close, high, and low values.
// Returns FastK, SlowK, and SlowD.
func (s *Stochastic) Update(close, high, low float64) (float64, float64, float64) {
	if math.IsNaN(close) || math.IsNaN(high) || math.IsNaN(low) {
		return math.NaN(), math.NaN(), math.NaN()
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	// Store high and low in circular buffer.
	s.highBuf[s.bufferIndex] = high
	s.lowBuf[s.bufferIndex] = low
	s.bufferIndex = (s.bufferIndex + 1) % s.fastKLength
	s.count++

	// Need at least fastKLength bars.
	if s.count < s.fastKLength {
		return s.fastK, s.slowK, s.slowD
	}

	// Find highest high and lowest low in the window.
	hh := s.highBuf[0]
	ll := s.lowBuf[0]

	for i := 1; i < s.fastKLength; i++ {
		if s.highBuf[i] > hh {
			hh = s.highBuf[i]
		}

		if s.lowBuf[i] < ll {
			ll = s.lowBuf[i]
		}
	}

	// Calculate Fast-K.
	diff := hh - ll
	if diff > 0 {
		s.fastK = 100 * (close - ll) / diff //nolint:mnd
	} else {
		s.fastK = 0
	}

	// Feed Fast-K to Slow-K smoother.
	s.slowK = s.slowKMA.Update(s.fastK)

	// Feed Slow-K to Slow-D smoother (only when Slow-K MA is primed).
	if s.slowKMA.IsPrimed() {
		s.slowD = s.slowDMA.Update(s.slowK)

		if !s.primed && s.slowDMA.IsPrimed() {
			s.primed = true
		}
	}

	return s.fastK, s.slowK, s.slowD
}

// UpdateScalar updates the indicator given the next scalar sample.
func (s *Stochastic) UpdateScalar(sample *entities.Scalar) core.Output {
	v := sample.Value
	fastK, slowK, slowD := s.Update(v, v, v)

	output := make([]any, 3) //nolint:mnd
	output[0] = entities.Scalar{Time: sample.Time, Value: fastK}
	output[1] = entities.Scalar{Time: sample.Time, Value: slowK}
	output[2] = entities.Scalar{Time: sample.Time, Value: slowD}

	return output
}

// UpdateBar updates the indicator given the next bar sample.
func (s *Stochastic) UpdateBar(sample *entities.Bar) core.Output {
	fastK, slowK, slowD := s.Update(sample.Close, sample.High, sample.Low)

	output := make([]any, 3) //nolint:mnd
	output[0] = entities.Scalar{Time: sample.Time, Value: fastK}
	output[1] = entities.Scalar{Time: sample.Time, Value: slowK}
	output[2] = entities.Scalar{Time: sample.Time, Value: slowD}

	return output
}

// UpdateQuote updates the indicator given the next quote sample.
func (s *Stochastic) UpdateQuote(sample *entities.Quote) core.Output {
	v := (sample.Bid + sample.Ask) / 2 //nolint:mnd

	return s.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

// UpdateTrade updates the indicator given the next trade sample.
func (s *Stochastic) UpdateTrade(sample *entities.Trade) core.Output {
	return s.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: sample.Price})
}
